### PR TITLE
Configuration page

### PR DESCRIPTION
**Pull request by mustiwebp**
Configuration-Page and the corresponding interaction (Creation of a patient) with the backend server have been implemented.

### CODE REVIEW

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                          content: Text(
+                            AppLocalizations.of(context)!.submitData,
+                          ),
+                        ),
+                      );
+                      controller.jumpToPage(_currentPage + 1);
+                    }
+                  } else {
+                    controller.jumpToPage(_currentPage + 1);
+                  }
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart:126]: That is a lot of unnecessary condition nesting, even with code duplication.

File: lib/l10n/app_de.arb [hunk 2]
```diff
@@ -76,5 +76,29 @@
...
+    "medicationAndTherapy": "Medikation & Therapie",
+    "infoText1": "Patient ID-12345 nimmt unter dem Abschnitt",
+    "infoText2": " an folgenden Angaben teil:",
+    "bodyAndMind": "Körper & Psyche",
+    "pain": "Schmerz",
+    "activityAndRest": "Aktivität & Ruhe",
+    "walkDistance": "Gehstrecke",
+    "sleepDuration": "Schlafdauer",
+    "sleepQuality": "Schlafqualität",
+    "vitalValues": "Vitalvalues",
```
Comment (S1rDan) [inline lib/l10n/app_de.arb:99]: Is english here intended?

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_activity_and_rest.dart [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+*    You should have received a copy of the GNU General Public License
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+
+import 'package:picos/util/backend_data.dart';
+
+/// shows page for configuration of "Activity & Rest"-information.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_activity_and_rest.dart:23]:
```suggestion
/// Shows page for configuration of "Activity & Rest"-information.
```
Typo

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_body_and_mind.dart [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+*
+*    You should have received a copy of the GNU General Public License
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:picos/util/backend_data.dart';
+
+/// shows page for configuration of "Body & Mind"-information.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_body_and_mind.dart:22]:
```suggestion
/// Shows page for configuration of "Body & Mind"-information.
```
Typo

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:queen_validators/queen_validators.dart';
+
+import 'package:picos/util/backend_data.dart';
+
+/// shows form for patient registration.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart:24]:
```suggestion
/// Shows form for patient registration.
```
Typo

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_medication_and_therapy.dart [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+*
+*    You should have received a copy of the GNU General Public License
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+import 'package:picos/util/backend_data.dart';
+
+/// shows page for configuration of "Medication & Therapy"-information.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_medication_and_therapy.dart:22]:
```suggestion
/// Shows page for configuration of "Medication & Therapy"-information.
```
Typo

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_summary.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+*    GNU General Public License for more details.
+*
+*    You should have received a copy of the GNU General Public License
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+
+/// shows summary page for configuration.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_summary.dart:21]:
```suggestion
/// Shows summary page for configuration.
```
Typo

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_vital_values.dart [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+*    You should have received a copy of the GNU General Public License
+*    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+
+import 'package:picos/util/backend_data.dart';
+
+/// shows page for configuration of vital values.
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_vital_values.dart:23]:
```suggestion
/// Shows page for configuration of vital values.
```
Typo

File: lib/util/backend_data.dart [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  /// contains patient's medication.
+  static bool medicationEnabled = false;
+
+  /// contains patient's therapy.
+  static bool therapyEnabled = false;
+
+  /// contains patient's doctor's visit.
+  static bool doctorsVisitEnabled = false;
+}
```
Comment (S1rDan) [inline lib/util/backend_data.dart:70]: As I understand, you are having here a class that only holds static values so you can use them inside the createPatient() method inside the Backend class.

The proper way of doing this would be to create a model and then using it, creating an object based on the model. For saving it the createPatient() method should take the Patient as a parameter.
If you have any difficulties with my proposal, let me know.

**Review by S1rDan — CHANGES_REQUESTED** [review]

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                }
+              },
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart:128]:
```suggestion
                if (_currentPage != _list.length - 1) {
                  if (_currentPage == 0 &&
                      formKeyConfiguration.currentState!.validate()) {
                    ScaffoldMessenger.of(context).showSnackBar(
                      SnackBar(
                        content: Text(
                          AppLocalizations.of(context)!.submitData,
                        ),
                      ),
                    );
                  }

                  controller.jumpToPage(_currentPage + 1);
                  return;
                }

                formKeyConfiguration.currentState!.save();
                bool backendCreatePatient = await Backend.createPatient();

                if (!backendCreatePatient) {
                  log('Fehler beim Anlegen des Patienten!');
                  return;
                }

                if (!mounted) {
                  return;
                }

                Navigator.of(context).pushNamed('/configuration-finish-screen');
              },
```

Suggestion how it could be clearer.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  const ConfigurationForm({Key? key}) : super(key: key);
+
+  @override
+  State<ConfigurationForm> createState() => _ConfigurationFormState();
+}
+
+class _ConfigurationFormState extends State<ConfigurationForm> {
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
```
[the remark this answers is not in the document]
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart:36]: > I am getting very confusing problems with the form. I can select both Mr and Mrs and cannot deselect it.
I will have a look on that.

> The way the "pages" are implemented may cause problems. E.g. you can just skip the form validation by scrolling down. And you are able to create "empty" users this way as long as you provide an e-mail adress.
Ehm, what do you mean by scrolling down? The form should be succesfully validated, if all fields are filled out in a plausible manner.

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            child: PicosInkWellButton(
+              text: AppLocalizations.of(context)!.proceed,
+              onTap: () async {
+                if (_currentPage == _list.length - 1) {
+                  formKeyConfiguration.currentState!.save();
+                  bool backendCreatePatient = await Backend.createPatient();
+                  if (backendCreatePatient) {
+                    if (!mounted) return;
+                    Navigator.of(context)
+                        .pushNamed('/configuration-finish-screen');
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart:108]:
```suggestion
                    Navigator.of(context)
                        .pushReplacementNamed('/configuration-finish-screen');
```
As far as I can see the previous screen isn't needed anymore, so why pushing on top of the stack?

File: lib/l10n/app_en.arb [hunk 13]
```diff
@@ -381,5 +381,125 @@
...
+        "description": "Medication & Therapy"
+    },
+
+    "infoText1": "Patient ID-12345 takes part in the section",
+    "@infoText1": {
+        "description": "Patient ID-12345 takes part in the section "
+    },
+
+    "infoText2": "with the following information",
+    "@infoText1": {
```
[the remark this answers is not in the document]
Comment (mustiwebp) [reply lib/l10n/app_en.arb:447]: Of course, it will be fixed.

File: lib/screens/study_nurse_screen/configuration_screen/configuration_screen.dart [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+                  fontSize: 25,
+                  color: theme.darkGreen3,
+                  fontWeight: FontWeight.bold,
+                ),
+              ),
+            ),
+            PicosInkWellButton(
+              text: AppLocalizations.of(context)!.letsstart,
+              onTap: () {
+                Navigator.pushNamed(context, '/configuration-pages');
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/configuration_screen.dart:72]:
```suggestion
                Navigator.pushReplacementNamed(context, '/configuration-pages');
```
As far as I can see the previous screen isn't needed anymore, so why pushing on top of the stack?

File: lib/l10n/app_de.arb [hunk 15]
```diff
@@ -76,5 +76,29 @@
     "achievedValues": "Heute erreichte Gesamteingabe",
     "days": "Tage",
     "month": "Monat",
-    "months": "Monate"
+    "months": "Monate",
+    "configuration": "Konfiguration",
+    "startConfiguration": "Konfiguration starten",
+    "letsstart": "Los geht's",
```
[the remark this answers is not in the document]
Comment (mustiwebp) [reply lib/l10n/app_de.arb:82]: Fixed!

File: lib/l10n/app_de.arb [hunk 16]
```diff
@@ -76,5 +76,29 @@
...
+    "infoText2": " an folgenden Angaben teil:",
+    "bodyAndMind": "Körper & Psyche",
+    "pain": "Schmerz",
+    "activityAndRest": "Aktivität & Ruhe",
+    "walkDistance": "Gehstrecke",
+    "sleepDuration": "Schlafdauer",
+    "sleepQuality": "Schlafqualität",
+    "vitalValues": "Vitalvalues",
+    "weightBMI": "Gewicht & BMI",
+    "heartFrequency": "Heart frequency",
```
[the remark this answers is not in the document]
Comment (mustiwebp) [reply lib/l10n/app_de.arb:101]: Fixed! English is not intended, of course.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_activity_and_rest.dart (hunk 3 above)
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/pages/configuration_activity_and_rest.dart:23]: Okay

File: lib/screens/study_nurse_screen/configuration_screen/configuration_finished_screen.dart [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+                  fontSize: 25,
+                  color: theme.darkGreen3,
+                  fontWeight: FontWeight.bold,
+                ),
+              ),
+            ),
+            PicosInkWellButton(
+              text: AppLocalizations.of(context)!.endConfiguration,
+              onTap: () {
+                Navigator.pushNamed(context, '/configuration-screen');
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/configuration_finished_screen.dart:72]:
```suggestion
                Navigator.of(context).pop();
```
The nurse or whoever is going to add new patients is likely going to have a screen preceding the configurationScreen. Just pop wherever you come from. As an alternative push the screen as replacement but don't create a tower of pisa.

**Review by S1rDan — CHANGES_REQUESTED** [review]
Additional things to change that I've noticed.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (S1rDan) [reply lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart:36]: I mean it literally. I am able to skip the validation provided by the button tap by scrolling to the "next page".

[the remark this answers is not in the document]
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart:36]: Okay, this will be fixed.

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart (hunk 10 above)
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart:128]: Your suggestion doesn't work unfortunately.

Comment (S1rDan) [reply lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart:128]: Yeah, I've noticed a mistake I made. But that should proof my point that this part of the code is hard to follow and has to be made clearer.
Btw. is there a particular reason why you are using a SnackBar displaying a message suggesting that the data has been transmitted? I mean, at this point nothing has been transmitted and it would give the user a false sense, telling him that.

File: lib/l10n/app_en.arb [hunk 18]
```diff
@@ -443,8 +443,8 @@
         "description": "Patient ID-12345 takes part in the section "
     },
 
-    "infoText2": "with the following information",
-    "@infoText1": {
+    "infoText2": " with the following information:",
```
Comment (S1rDan) [inline lib/l10n/app_en.arb:446]: You have added a second (redundant) space.

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart (hunk 1 above)
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart]: This should be fixed now.

File: lib/l10n/app_de.arb (hunk 2 above)
Comment (mustiwebp) [reply lib/l10n/app_de.arb]: > Is english here intended?

No, it isn't. Fixed!

File: lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart (hunk 10 above)
Comment (mustiwebp) [reply lib/screens/study_nurse_screen/configuration_screen/configuration_pages.dart]: Yeah thanks, it works now.

**Review by S1rDan — COMMENTED** [review]
These are looking good so far.

File: lib/util/backend_data.dart [hunk 19]
```diff
@@ -0,0 +1,70 @@
+/// enumeration for gender-variable.
+enum Gender {
+  /// element for denoting the title of men
+  male,
+
+  /// element for denoting the title of women
+  female,
+}
+
```
Comment (inchw0rm) [inline lib/util/backend_data.dart:9]: Please include at least "other"

Comment (S1rDan) [reply lib/util/backend_data.dart:9]: Another option would be to name the enum "Sex", because other than genders there can only be two or three sexes. Also I think that the biological constitution is in a medical application more relevant rather than what the person identifies with.

Comment (mustiwebp) [reply lib/util/backend_data.dart]: Here, I added the option "diverse".

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_body_and_mind.dart [hunk 20]
```diff
@@ -17,20 +17,29 @@
...
   State<ConfigurationBodyAndMind> createState() =>
       _ConfigurationBodyAndMindState();
 }
 
 class _ConfigurationBodyAndMindState extends State<ConfigurationBodyAndMind> {
+  /// Local variable for pain.
+  bool entryPainEnabled = false;
+
+  /// Local variable for blood sugar levels.
+  bool entryPhq4Enabled = false;
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_body_and_mind.dart:41]: Could be private.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart [hunk 21]
```diff
@@ -17,21 +17,27 @@
...
+  /// Callback function for form.
+  final Function(String, String) callbackForm;
 
   @override
   State<ConfigurationForm> createState() => _ConfigurationFormState();
 }
 
 class _ConfigurationFormState extends State<ConfigurationForm> {
+  /// Local variable for form of address.
+  FormOfAddress entryFormOfAddress = FormOfAddress.female;
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_form.dart:39]: Can be private.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_medication_and_therapy.dart [hunk 22]
```diff
@@ -33,6 +38,15 @@ class ConfigurationMedicationAndTherapy extends StatefulWidget {
...
 class _ConfigurationMedicationAndTherapyState
     extends State<ConfigurationMedicationAndTherapy> {
+  /// Local variable for medication.
+  bool entryMedicationEnabled = false;
+
+  /// Local variable for therapy.
+  bool entryTherapyEnabled = false;
+
+  /// Local variable for doctor visits.
+  bool entryDoctorsVisitEnabled = false;
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_medication_and_therapy.dart:48]: Can be private.

File: lib/screens/study_nurse_screen/configuration_screen/pages/configuration_vital_values.dart [hunk 23]
```diff
@@ -18,20 +18,35 @@
...
+  bool entryWeightBMIEnabled = false;
+
+  /// Local variable for heart frequency.
+  bool entryHeartFrequencyEnabled = false;
+
+  /// Local variable for blood pressure.
+  bool entryBloodPressureEnabled = false;
+
+  /// Local variable for blood sugar levels.
+  bool entryBloodSugarLevelsEnabled = false;
```
Comment (S1rDan) [inline lib/screens/study_nurse_screen/configuration_screen/pages/configuration_vital_values.dart:48]: Can be private.

**Review by S1rDan — CHANGES_REQUESTED** [review]

File: lib/util/backend_data.dart (hunk 19 above)
Comment (mustiwebp) [reply lib/util/backend_data.dart]: Finally, I have decided to leave it as a binary option as shown by the mockup. We should have a discussion with the other Stakeholders.

**Comment by S1rDan** [comment]
> I'm about to do some tests. Then it should be finished and ready to merge from my side.

You have forgot to take care of _patient_profile.dart_ as I noted here: [https://github.com/healthcare-it-solutions/picos/pull/23#discussion_r991910000](url)

**Review by S1rDan — APPROVED** [review]

**Comment by S1rDan** [comment]
Can be merged now imo.

File: pubspec.yaml [hunk 24]
```diff
@@ -41,6 +41,7 @@ dependencies:
   fl_chart: ^0.55.1
   parse_server_sdk: ^3.1.2
   queen_validators: ^1.0.0
+  random_password_generator: ^0.2.1
```
Comment (inchw0rm) [inline pubspec.yaml:44]: May I ask why? This change hasn't been discussed.

File: lib/routes.dart [hunk 25]
```diff
@@ -39,6 +43,10 @@ class Routes {
       '/family-members': (BuildContext ctx) => const FamilyMembersScreen(),
       '/mainscreen': (BuildContext ctx) => const BottomBar(title: 'PICOS',),
       '/studynursescreen': (BuildContext ctx) => const StudyNurseScreen(),
+      '/configuration-screen': (BuildContext ctx) => const ConfigurationScreen(),
+      '/configuration-pages': (BuildContext ctx) => const ConfigurationPages(),
+      '/configuration-finish-screen': (BuildContext ctx) => const ConfigurationFinishedScreen(),
+      '/overview': (BuildContext ctx) => const OverviewScreen(),
     };
   }
```
Comment (inchw0rm) [inline lib/routes.dart:51]: Please keep in mind that we shouldn't use a flat route structure, please let's have a tree like structure and introduce route names like `/studynursescreen/configuration-page`

**Review by inchw0rm — COMMENTED** [review]
Looks fine to me except for the random_passwords I asked to comment on.

**Comment by inchw0rm** [comment]
Please remove random_password_generator

File: lib/routes.dart (hunk 25 above)
Comment (mustiwebp) [reply lib/routes.dart]: I've implemented your suggestion. You may have a look :)

**Comment by mustiwebp** [comment]
> Please remove random_password_generator
Should there be any substitutions?

**Comment by inchw0rm** [comment]
As long as it's not a library with 2 stars and a last updated more than a year ago, you can pick any.

**Comment by mustiwebp** [comment]
> As long as it's not a library with 2 stars and a last updated more than a year ago, you can pick any.

I substituted it by a straightforward string generator. Hence, the package has been removed now.

**Comment by inchw0rm** [comment]
Looks good, can I merge now?
